File: app/src/main/java/com/example/sachinsandbhor/retrofitsample/presentation/movieList/MovieListModel.java
package com.example.sachinsandbhor.retrofitsample.presentation.movieList;

import com.example.sachinsandbhor.retrofitsample.BuildConfig;
import com.example.sachinsandbhor.retrofitsample.app.MovieService;
import com.example.sachinsandbhor.retrofitsample.model.MovieResponse;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Sachin.Sandbhor on 06-02-2018.
 */

public class MovieListModel implements MovieListContract.MovieListModel{

    MovieListContract.apiServices apiServices;

    MovieListContract.MovieListModel.MovieListPresenterCallBack movieListPresenterCallBack;

    public MovieListModel(MovieListContract.MovieListModel.MovieListPresenterCallBack movieListPresenterCallBack) {
        apiServices = MovieService.getClient().create(MovieListContract.apiServices.class);
        this.movieListPresenterCallBack = movieListPresenterCallBack;
    }

    @Override
    public void getMovieList() {
        Observable<MovieResponse> movieResponseObservable = apiServices.getRatedMovies(BuildConfig.API_KEY)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());

        movieResponseObservable.subscribe(new Observer<MovieResponse>() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onNext(MovieResponse movieResponse) {
                movieListPresenterCallBack.onMovieListReceived(movieResponse);
            }

            @Override
            public void onError(Throwable e) {
                movieListPresenterCallBack.onError(e.getMessage());
            }

            @Override
            public void onComplete() {

            }
        });
    }
}

File: retrofitkotlin/src/main/java/com/sachinsandbhor/retrofitkotlin/presentation/movielist/MovieListModel.kt
package com.sachinsandbhor.retrofitkotlin.presentation.movielist

import com.sachinsandbhor.retrofitkotlin.BuildConfig
import com.sachinsandbhor.retrofitkotlin.domain.MovieListResponse
import com.sachinsandbhor.retrofitkotlin.network.MovieService
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers


/**
 * Created by Sachin.Sandbhor on 08-03-2018.
 */
class MovieListModel(var movieListModelCallBack: MovieListContract.MovieListModel.MovieListCallback) : MovieListContract.MovieListModel {

    lateinit var apiService:MovieListContract.apiService

    init {
         apiService = MovieService.getClient()
                .create(MovieListContract.apiService::class.java)
    }

    override fun getMovieList(pageNo:Int) {
        val movieResponseObservable = apiService.getRatedMovies(BuildConfig.API_KEY, pageNo)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())

        movieResponseObservable.subscribe(object : Observer<MovieListResponse> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onNext(movieResponse: MovieListResponse) {
                movieListModelCallBack.movieListReceived(movieResponse)
            }

            override fun onError(e: Throwable) {
                movieListModelCallBack.onError(e.message)
            }

            override fun onComplete() {

            }
        })
    }
}